swarm-derive/src/lib.rs

```diff
@@ -757,6 +757,7 @@ fn build_struct(ast: &DeriveInput, data_struct: &DataStruct) -> syn::Result<Toke
                 std::task::Poll::Ready(#network_behaviour_action::CloseConnection { peer_id, connection }) => {
                     return std::task::Poll::Ready(#network_behaviour_action::CloseConnection { peer_id, connection });
                 }
+                std::task::Poll::Ready(_) => {},
```

Comment:
So this is tricky right. This should never happen because we should _always_ update this crate to generate all possible match arms. As a result, I think we should use `unreachable!` here. If we hit this, this is a bug.

Reply:
I think we can actually cleanly solve this using `ToSwarm::map_in`. That function is defined within `libp2p-swarm` and will those correctly map all variants. It should be something like:

And instead of `generate_event_match_arm`, we can probably use `map_out`.

So it should be something like:

```rust
#trait_to_impl::poll(&mut self.#field, cx).map_out(|event| #into_out_event).map_in(|event| #wrapped_event)
```

Perhaps even cleaner would be to generate the closures directly because then the variable names are co-located.

Reply:
That is a neat solution leveraging that we can exhaustively match in `libp2p-swarm`.